### ISSUE TITLE
Cannot do checks inside a thread

### ISSUE THREAD

**Issue by cschreib** [snitch-org/snitch#188]
Initially reported [here](https://github.com/snitch-org/snitch/issues/80#issuecomment-2126174211).

We currently rely on thread-local state to link the checks to the currently-running test. This state is set by the registry when starting the test, and users cannot set it. Therefore, if a test is launching a thread and trying to run checks inside the thread, the test state is not set and check functions terminate.

```c++
TEST_CASE("foo") {
    std::size_t i = 0;
    std::jthread t([&] {
        CHECK(i == 0); // std::terminate called
    });
}
```

If we want to allow this, we need to somehow fill in the test state in the child thread. Ideally this would be automatic, for example: if the test state is empty, get the state of the parent thread, repeat until a state is found (and terminate if no parent thread). But I don't think this is possible, because standard threads have no concept of a "parent" thread.

Also, consider this example of a basic thread pool, in which the worker threads are lazily created when the first task is pushed to the pool:
```c++
class ThreadPool {
    std::vector<std::jthread> threads;
    std::vector<std::function<void()>> tasks;

public:
    template<typename F>
    void run(F&& func) {
        if (threads.empty()) {
            for (std::size_t i = 0; i < 8; ++i) {
                threads.emplace_back([&]() { /* process tasks */ });
            }
        }

        tasks.push_back(std::forward<F>(func));
    }
} pool;

TEST_CASE("test 1") {
    pool.run([]() {
        CHECK(1 == 1);
    });
}

TEST_CASE("test 2") {
    pool.run([]() {
        CHECK(1 == 1);
    });
}
```

The worker threads are created in `"test 1"`, and thus could inherit the test state from `"test 1"`. So far so good. But then consider the following problems:
 1. The thread pool had a long queue to process, so it only gets around to running the task after `"test 1"` was finished; there's no running test at that point, so we should be careful not to use dangling state from `"test 1"` (correct answer here would be to terminate, as is done today).
 2. `"test 1"` finishes and `"test 2"` starts; the worker threads created by `"test 1"` are still alive, but the correct test state to use is now that of `"test 2"`. How can we notify the threads of that?
 3. Consider further that `"test 1"` and `"test 2"` may be [run in parallel in the future](https://github.com/snitch-org/snitch/issues/33), so there's no way to predict which of the two tests is going to initialise the worker threads, and if each test is pushing a number of tasks on the pool, the correct test state to use could have to go back and forth between `"test 1"` and `"test 2"` for the same worker thread.

I need to think about this some more, but currently I don't think there's a way to get this to work automatically (and do the correct, unsurprising thing). My thinking at the moment is that this would require an explicit forwarding of the test state, something like:
```c++
TEST_CASE("test 1") {
    auto& test_state = snitch::impl::get_current_test();
    pool.run([&]() {
        snitch::impl::set_current_test(&test_state);
        CHECK(1 == 1);
    });
}
```

But that doesn't solve problem 1 above (dangling test state pointer).

Maybe I should first read how other test frameworks approach this.

**Comment by cschreib** [comment]
From what I can piece together:

In *Catch2* the equivalent to our `get_current_test()` is called `getResultCapture()`, which in turn calls `getCurrentContext().getResultCapture()`. This current context is a `static` class member (lazily initialised), hence it is shared by all threads. This assumes there's only one test running at any given time, [which is a design decision of Catch2](https://github.com/catchorg/Catch2/issues/1308). So, in *Catch2*, parallel test execution is simply not supported.

Same in *doctest*; there is a global `ContextState` (`g_cs`); parallel test execution is [also not supported](https://github.com/doctest/doctest/issues/168).

Same for *GoogleTest* with `UnitTest::GetInstance()` which returns a function `static` variable; parallel test execution is [also not supported](https://github.com/google/googletest/issues/465).

In all these testing frameworks the recommended approach is to use sharding (i.e., run a different subset of the tests in separate processes). This solves the problem, though it is not without downsides (more overhead in process launch, inconvenient when running tests by hand on the command line -- probably requires a runner script to wrap the launches and collate the results).

*boost-ut* has a [parallel test runner example](https://github.com/boost-ext/ut/blob/master/example/cfg/parallel_runner.cpp). The test runner is configured as a global variable `ut::cfg`, which is accessed to process all events. Issue is, I don't see how the assertions are linked to the currently running test; it seems they are not. The default reporter is fully sequential, and expects events to be generated in the order `test_begin`, `assertion_fail`, `test_end`; hence it relies on event ordering to identify the test that triggered the assertion. In the parallel test runner examples, the events are processed by the runner and forward to the default reporter. This won't work for parallel reporting, unless I'm missing something :thinking: I don't think I am: the example runs fine at first sight, but only because the execution isn't actually parallel (GDB shows no thread being created). If implemented explicitly with threads, one for each suite, the output I get is indeed incorrect:
```
test.1.1
test.2.1
test.2.2
test.1.2
test.2.3
test.1.3
Running "test.1
 "test.2.1"...1
  parallel_runner.cpp:64:FAILED [
 "test.1.2"...2 == 1]parallel_runner.cpp:55:FAILED [2 == 1]
FAILED
Running "test.2.3"...PASSED
Running "test.1.3"...PASSED

===============================================================================
tests:   3 | 1 failed
asserts: 6 | 4 passed | 2 failed
.
```

**Comment by cschreib** [comment]
The solution adopted in *Catch2*/*doctest*/*GoogleTest* (global test instance) solves the problem of having to propagate the currently running test to child threads. However, it still has issues with asynchronous checks.

In the examples below, all of these frameworks will report an assertion in the wrong test (the check inside the thread is reported in the second test case instead of the first), and will segfault if an assertion is reported in a child thread when no test is running (e.g., if commenting out the second test case).

*Catch* / *doctest*:
```c++
#include "catch2/catch_all.hpp"
#include <thread>

TEST_CASE("testing 1") {
    new std::thread([] {
        std::this_thread::sleep_for(std::chrono::seconds(1));
        CHECK(1 == 2);
    });
}

TEST_CASE("testing 2") {
    std::this_thread::sleep_for(std::chrono::seconds(2));
}
```

*GoogleTest*:
```c++
#include "gtest/gtest.h"
#include <thread>

TEST(TestSuite, Testing1) {
    new std::thread([] {
        std::this_thread::sleep_for(std::chrono::seconds(1));
        ASSERT_EQ(1, 2);
    });
}

TEST(TestSuite, Testing2) {
    std::this_thread::sleep_for(std::chrono::seconds(2));
}
```

**Comment by cschreib** [comment]
It's also worth noting that `REQUIRE(...)` checks, which are meant to immediately stop the test on failure, are inherently difficult to use within a separate thread. It would require a `try`/`catch` in the thread to catch and handle the abort exception, and it would then need to propagate the abort message to the parent thread. That's possible, but not easy. The simplest might be to stick to the default C++ behavior for unhandled exceptions in threads, which simply terminates. This isn't as nice as a `REQUIRE(...)` outside a thread, which only stops the current test. But it's the same behavior one would get with exceptions disabled (`REQUIRE(...)` terminates immediately).

**Comment by cschreib** [comment]
Another problem: what about `CAPTURE` and `SECTION`? With the current architecture, these are stored alongside the test state, and would be shared by all threads pointing to that test. This is probably wrong, I think we'd want the captures and sections to remain thread-local all the time. Otherwise, a check executed in a child thread would be reported with whatever section and captures are currently active in the parent thread, which probably doesn't make sense.

**Comment by cschreib** [comment]
> But maybe we don't need the full `shared_ptr` interface; a raw pointer to a stack object and a single atomic ref count would be sufficient (we don't need weak references, `release()`, `make_shared()`, etc.).

This could work:
```c++
#include <atomic>

using ref_count_t = std::int_fast16_t; // we should make that configurable

template<typename T>
class shared_ref;

template<typename T>
class ref_counted {
public:
    T data;

    ref_counted() = default;

    template<typename ... Args>
    ref_counted(Args&& ... args) : data(std::forward<Args>(args)...) {}

    ~ref_counted() {
        wait_until_exclusive();
    }

private:
    std::atomic<ref_count_t> refs = 1;

    void push_ref() noexcept {
        refs += 1; // open question: assert on overflow? Would make most of the interface not noexcept
    }

    void pop_ref() noexcept {
        refs -= 1; // open question: assert on undeflow?
        refs.notify_one(); // C++20; would have to check compiler support
    }

    template<typename U>
    friend class shared_ref;

public:
    shared_ref<T> get_ref() noexcept;

    void wait_until_exclusive() const noexcept {
        while (true) {
            const auto current_ref = refs.load();
            if (current_ref <= 1) {
                break;
            }

            refs.wait(current_ref); // C++20; would have to check compiler support
        }
    }
};

template<typename T>
class shared_ref {
    ref_counted<T>* ptr = nullptr;

    template<typename U>
    friend class ref_counted;

    shared_ref(ref_counted<T>* p) noexcept : ptr(p) {}

public:
    shared_ref() = default;

    shared_ref(const shared_ref& other) noexcept :
        ptr(other.ptr != nullptr ? other.ptr->get_ref().ptr : nullptr) {}

    shared_ref(shared_ref&& other) noexcept :
        ptr(other.ptr) { other.ptr = nullptr; }

    ~shared_ref() {
        reset();
    }

    shared_ref& operator=(const shared_ref& other) noexcept {
        reset();
        ptr = other.ptr != nullptr ? other.ptr->get_ref() : nullptr;
        return *this;
    }

    shared_ref& operator=(shared_ref&& other) noexcept {
        reset();
        std::swap(ptr, other.ptr);
        return *this;
    }

    void reset() noexcept {
        if (ptr != nullptr) {
            ptr->pop_ref();
            ptr = nullptr;
        }
    }

    T* operator->() const noexcept {
        return &ptr->data;
    }

    T& operator*() const noexcept {
        return ptr->data;
    }
};

template<typename T>
shared_ref<T> ref_counted<T>::get_ref() noexcept {
    push_ref();
    return this;
}
```

Usage: 
 - `ref_counted<int> parent(42);` creates a new ref-counted object on the stack. No heap allocation.
 - `parent.get_ref()` creates a new reference, bumping the ref count by 1.
 - When that reference goes out of scope or is `reset()`, the ref count goes down by 1.
 - `parent.wait_until_exclusive();` waits until it holds the only reference to the object. At that point the object is safe to destroy. This is automatically called in the destructor of `parent`, to ensure no other ref exists before the object is destroyed.
